library/ssl_tls13_client.c

```diff
@@ -2174,24 +2174,8 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
          * client MUST verify that the server's selected_identity
          * is 0. If any other value is returned, the client MUST
          * abort the handshake with an "illegal_parameter" alert.
-         *
-         * RFC 8446 4.2.10
-         * In order to accept early data, the server MUST have accepted a PSK
-         * cipher suite and selected the first key offered in the client's
-         * "pre_shared_key" extension. In addition, it MUST verify that the
-         * following values are the same as those associated with the
-         * selected PSK:
-         * - The TLS version number
-         * - The selected cipher suite
-         * - The selected ALPN [RFC7301] protocol, if any
-         *
-         * We check here that when early data is involved the server
-         * selected the cipher suite associated to the pre-shared key
-         * as it must have.
          */
-        if (handshake->selected_identity != 0 ||
```

Comment:
When parsing the ServerHello we do not know if the server will accept the early data or not thus at this point it is valid for the server to select the first proposed pre-shared key and a ciphersuite that is different from the ciphersuite associated to the selected pre-shared key (only the associated hash has to match at this point, but maybe that currently we accept only the same ciphersuite and that would rather be there that we should improve). Then here we know that the server accepts the early data and it makes sense to me to check the ciphersuite.

Reply:
Agree. I just think about that.  We should check ciphersuite also. But we should check it with proposed ciphersuite. 

And I just checked `mbedtls_ssl_tls13_compute_handshake_transform()` , It is computed base on `handshake->ciphersuite_info`. So the first commit does make sense. 

@yanrayw , please revert to the first commit with some comments

Reply:
Seems we should revert the title of this PR also. `"TLS13: CLI: EarlyData: Remove unnecessary ciphersuite in EE "`